docs/configuration/http_filters/lua_filter.rst

```diff
@@ -0,0 +1,325 @@
+.. _config_http_filters_lua:
+
+Lua
+===
+
+.. attention::
+
+  The Lua scripting HTTP filter is **experimental**. Use in production at your own risk. It is
+  being released for initial feedback on the exposed API and for further development, testing,
+  and verification. This warning will be removed when we feel that the filter has received enough
+  testing and API stability to call it generally production ready.
+
+Overview
+--------
+
+The HTTP Lua filter allows `Lua <https://www.lua.org/>`_ scripts to be run during both the request
+and response flows. `LuaJIT <http://luajit.org/>`_ is used as the runtime. Because of this, the
+supported Lua version is mostly 5.1 with some 5.2 features. See the `LuaJIT documentation
+<http://luajit.org/extensions.html>`_ for more details.
+
+The filter only supports loading Lua code in-line in the configuration. If local filesystem code
+is desired, a trivial in-line script can be used to load the rest of the code from the local
+environment.
+
+The design of the filter and Lua support at a high level is as follows:
+
+* All Lua environments are :ref:`per worker thread <arch_overview_threading>`. This means that
+  there is no truly global data. Any globals create and populated at load time will be visible
+  from each worker thread in isolation. True global support may be added via an API in the future.
+* All scripts are run as coroutines. This means that they are written in a synchronous style even
+  though they may perform complex asynchronous tasks. This makes the scripts substantially easier
+  to write. All network/async processing is performed by Envoy via a set of APIs. Envoy will
+  yield the script as appropriate and resume it when async tasks are complete.
+* **Do not perform blocking operations from scripts.** It is critical for performance that
+  Envoy APIs are used for all IO.
+
+Currently supported high level features
+---------------------------------------
+
+**NOTE:** It is expected that this list will expand over time as the filter is used in production.
+The API surface has been kept small on purpose. The goal is to make scripts extremely simple and
+safe to write. Very complex or high performance use cases are assumed to use the native C++ filter
+API.
+
+* Inspection of headers, body, and trailers while streaming in either the request flow, response
+  flow, or both.
+* Modification of headers and trailers.
+* Blocking and buffering the full request/response body for inspection.
+* Performing an outbound async HTTP call to an upstream host. Such a call can be performed while
+  buffering body data so that when the call completes upstream headers can be modified.
+* Performing a direct response and skipping further filter iteration. For example, a script
+  could make an upstream HTTP call for authentication, and then directly respond with a 403
+  response code.
```

Comment:
Can sleeps be used? Or timer routines? Coz that will allow doing custom fault injection scripts. If not, suggest mentioning this. Use of timers is common in OpenResty world.

Reply:
I will add an msleep() method as a follow up when I break apart the PR. It will be trivial to add.